### ISSUE TITLE
Binary Cross Entropy loss is incorrect: author confused about what BCE is

### ISSUE THREAD

**Issue by mcourteaux** [keras-team/keras#17766]
**Describe the problem**.

The Binary Cross Entropy loss implementation computes the mean (along an axis of choice) of the binary cross entropy loss. Obviously, this mean is pointless, and has caused a lot of confusion with the people writing the documentation it seems.

The mean reduction I'm talking about is at the last line of this function (L2155):
https://github.com/keras-team/keras/blob/f9336cc5114b4a9429a242deb264b707379646b7/keras/losses.py#L2116-L2158

So clearly, judging by its name, this function promises to calculate the _binary crossentropy_. However, the reduction axis there is very weird, and the author clearly got confused with the more general crossentropy definition (where you indeed take the expected value of `-log(x)`. In other words, with a summation, you take the sum over `-x*log(x)`, with `x` being the probabilities summing to 1), like here (from [Wiki](https://en.wikipedia.org/wiki/Cross_entropy#Definition)):
![image](https://user-images.githubusercontent.com/845012/229758225-30d1ed6c-9905-4f72-be07-f17095570786.png)

However, the mistake is that the `backend.binary_crossentropy` function already performs this reduction internally:

https://github.com/keras-team/keras/blob/f9336cc5114b4a9429a242deb264b707379646b7/keras/backend.py#L5687-L5717

The summation is in the fact that the implementation computes `bce` with the two terms (one in `bce = ...` at L5715, and one in `bce += ...` at L5716). These two terms are the summation:
![image](https://user-images.githubusercontent.com/845012/229758225-30d1ed6c-9905-4f72-be07-f17095570786.png)

So, as an intermediate conclusion: `keras.backend.binary_crossentropy` has the correct implementation and does what it promises.

Now, the documentation of `keras.metrics.binary_crossentropy()` is suggesting to use it in a pointless way:

https://github.com/keras-team/keras/blob/f9336cc5114b4a9429a242deb264b707379646b7/keras/losses.py#L2121-L2126

So, clearly, the writer of the function and documentation thought there was a need to supply the full list of probabilities that sum to 1. All pairs in this demonstration sum to one (0.6 + 0.4, and 0 + 1). Then, the `backend.binary_crossentropy()` is called on these inputs, which computes 4 (four!) binary crossentropies, which has a resulting shape of `(2, 2)` (just like the input shape!), which is then mean-reduced over the last axis to finally yield `(2,)` as a shape.

Now clearly, the above described example from the documentation is just a demonstration of how to use multi-class crossentropy, applied on a case with only 2 classes. Under the hood, the function calculates the binary crossentropy twice for every input, and is then averaged out. The final output of the example is correct, but the whole function is completely bananas.

**Describe the current behavior**.

The function does not correctly compute binary cross-entropy if you do not provide it with pairs of probabilities that sum to 1.

**Describe the expected behavior**.

It just computes the actual binary cross-entropy, and the mean-reduction is thrown out.

A correct documentation example would be:

    Standalone usage:
    >>> y_true = [1, 0]
    >>> y_pred = [0.6, 0.4]
    >>> loss = tf.keras.losses.binary_crossentropy(y_true, y_pred) # No axis nonsense
    >>> assert loss.shape == (2,)

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): yes
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing): remove the backend.mean call, remove the axis parameter. This probably severely breaks whatever it is that people have built around this mistake (either depend on it, or a workaround), but long-term, it's the right thing to do.

**Comment by Frightera** [comment]
Hi @mcourteaux,

> The function does not correctly compute binary cross-entropy if you do not provide it with pairs of probabilities that sum to 1.

Can you provide an example for that? Normally, it expects probs to sum up to 1 unless `from_logits` is set to `True`.

Indeed that `backend.mean` is confusing, it is used to support `sample_weight` when needed. As far as I understand it makes sure of that loss is computed correctly when `sample_weight` is provided, should not really matter when using normal Tensors, but can be a problem (without `backend.mean()`) when using `RaggedTensor`. 

Specifically, loss functions in Keras expects 2D inputs at least, which is `[batch, d0]`, but can have `N` dimensional inputs aswell (including BCE). This is also mentioned in [base Loss class' call method](https://www.tensorflow.org/api_docs/python/tf/keras/losses/Loss#__call__). Maybe a team member can also confirm this.

However I did not see any inconsistencies in calculated values, they return the expected results everytime.

I agree with that the documentation can be more clear.

**Comment by mcourteaux** [comment]
I was writing an extensive reply. I did not forget about this. I will try to finish this today. Thanks for looking into this. :smile:

**Comment by mcourteaux** [comment]
I see how the set of "rules" around a loss _always_ removing the last dimension somewhat conveniently works. The main reason I see, is that a binary classification network would end with a `Dense(1, activation='sigmoid')` layer, which causes the shape of the output to be `(batch, 1)` instead of `(batch,)`. This is where the `backend.mean()` aspect of the binary cross-entropy conveniently removes the last dimension (which was pointless to begin with).

In a more general or flexible case, we would like to provide sample weights which specify also weights of the last dimension. This is often true for multi-class multi-label problems. Example: _which of the K animal species are present in a picture?_ Obviously, you want the network to output multiple labels for one picture, and the loss has to be the _binary_ cross-entropy loss evaluated on every individual output. Intuitively, one would design a network architecture that would end in a `Dense(K, activation='sigmoid')` layer, having the output shape to be `(batch, K)`.

Now, when trying to specify sample weights of the shape `(batch, K)` to accommodate uncertainties on labels and have the (animal) class weights be integrated, this becomes impossible as the basic loss function already averages out over the last dimension over the K classes, so there is no chance anymore to apply class weights.

(There is another reason I believe why class-weights are broken in this case: it is expected to be a dictionary instead of a flat array. I don't think this dictionary works with the multi-label aspect of this example.)

So to be able to use class-weights, you need to do one of two things:

 - Use a custom loss function which doesn't to the unfortunate mean-reduction over the last dimension.
 - Use an extra Lambda() layer as a workaround to add an extra dimension of length 1 to the output, which then can be reduced away again by the loss function. This breaks saving/restoring models.

---

Thinking of all of this, I feel like the basic assumption that a loss function _MUST_ reduce at least one dimension is pointless; and as far as I understand, it is not required to "make the sample weights work", as the documentation states that broadcasting is also possible:

> If the shape of sample_weight is [batch_size, d0, .. dN-1] **(or can be broadcasted to this shape)**, then each loss element of y_pred is scaled by the corresponding value of sample_weight. (Note on dN-1: all loss functions reduce by 1 dimension, usually axis=-1.)

I don't see the problem with having a loss function not reduce the last dimension. You can still broadcast any shape `[batch_size, d0, .. dL]` (with `L` <= `N`) broadcast to `[batch_size, d0, ..., dN]`.

In case of a simple binary classification network, you'd have as output of the `Dense(1)` layer `(batch, 1)`, which the loss function applies the BCE to, and you simply get again `(batch, 1)`, which the sample weights correctly can be mapped to as `(batch,)` can be broadcast to `(batch, 1)`, which _then_ can be reduced to `(batch,)` before accumulating into the Reduction statistics. The part after the italic "then", already happens today in Keras, AFAIK.

---
 
> > The function does not correctly compute binary cross-entropy if you do not provide it with pairs of probabilities that sum to 1.
> 
> Can you provide an example for that? This might be related to from_logits parameter.

No, I was wrong: my bad. It does compute it correctly. It just calculates it twice, obtains identical results, and then the `backend.mean()` takes the average of the two identical numbers, to again yield the same (correct) result.

**Comment by Frightera** [comment]
@mcourteaux 

> Now, when trying to specify sample weights of the shape (batch, K) to accommodate uncertainties on labels and have the (animal) class weights be integrated, this becomes impossible as the basic loss function already averages out over the last dimension over the K classes, so there is no chance anymore to apply class weights.

Fair concerns on `class_weights` while using `sample_weight` at the same time, I don't think Keras supports `class_weights` properly when the task is multilabel classification, it's just not implemented. Actually `class_weights` are just multiplied with `sample_weights` (if provided) and put into a `tf.data.Dataset` as the 3rd element (before even training starts), then this `sample_weights` are just passed into the loss function. Before applying the `sample_weights`, there is some kind of broadcasting:

https://github.com/keras-team/keras/blob/4d1ca982a64b86c1ac3a113f66c4f17986fba243/keras/utils/losses_utils.py#L343-L358

> So to be able to use class-weights, you need to do one of two things:

> 1) Use a custom loss function which doesn't to the unfortunate mean-reduction over the last dimension.
> 2) Use an extra Lambda() layer as a workaround to add an extra dimension of length 1 to the output, which then can be reduced away again by the loss function. This breaks saving/restoring models.

I think in the multi-label case, you need to use a custom loss to do the weighting manually.

---

> Thinking of all of this, I feel like the basic assumption that a loss function MUST reduce at least one dimension is pointless.

This reduction is required in the case of `RaggedTensors`. Consider the case:

```python
y_true = tf.ragged.constant([[1, 0, 1], [0]])
y_pred = tf.ragged.constant([[1, 1, 1], [0]], dtype=tf.float32)
```

In order to calculate the loss using `RaggedTensor`, the loss function is wrapped into another function:
https://github.com/keras-team/keras/blob/f0525e5c07f443e038f0d41e90691548c21d5228/keras/losses.py#L2464-L2470

Operations on `RaggedTensors` are kind of different, AFAIK it calls the loss function twice (happens in TensorFlow backend). If this reduction is not performed, you'll end up having errors saying that shapes do not match with the expected ones. That is because loss is called twice in this case and reduction is applied independently on each dimension. So it expects `shape = ()` instead of `shape = (3, )` for the first dimension. After all you can easily do any calculations when you have shapes as `( )`, basically a scalar value.

That's why I said _it is used to support sample_weight when needed_, now `sample_weight` can be correctly applied to those scalar values.

Yes, this should not effect when tensors with uniform length, this applies to the `RaggedTensors` only. Hope this clarifies your query, that's what I know about this reduction thing.

**Comment by adamvvu** [comment]
A convenient result due to the extra mean reduction however is if we have an output of the shape (batch, K), e.g. a multi-label problem/vector-valued outputs with something like `Dense(K, activation='sigmoid')` and `from_logits=False`. For example,
```
y_true = [[1, 0, 1], [0, 1, 0]]
```
Then the current implementation would automatically compute/average the BCE for each column. 

Without it, we would have to specify a multi-output model each with `Dense(1, activation='sigmoid')` which is not the same thing. 

But I have encountered similar issues with trying to apply sample weights of shape `(batch, K)`.

**Comment by matangover** [comment]
@mcourteaux I think you are rightly confused, but it's just a misunderstanding. I was confused myself but I think I understand it now.

`binary_crossentropy` computes element-wise BCE (i.e. each element in y_true/y_pred is an independent binary classification problem) and then averages the loss over the last dimension (i.e. the last dimension represents multiple classification problems that should all be taken into account).

This seems to be intended, and matches e.g. a multi-class multi-label problem:
```python
y_true = [[0,   0,   1  ], [1,   1,   0  ]]
y_pred = [[0.0, 0.1, 0.8], [0.7, 0.8, 0.2]]
```
The loss is then calculated sample-wise (one result for each sample in the batch). This is consistent with how all loss functions work.
```python
binary_crossentropy(y_true, y_pred)
# [0.154, 0.228]
```

Nonetheless, you can disable the averaging:
```python
binary_crossentropy(y_true, y_pred, axis=[])
# [0.000 , 0.356, 0.105],
# [0.356 , 0.105, 0.223]]
```

Bottom line, if you are using `binary_crossentropy` to "compute the same number twice", I think you're not using it as intended.

**Comment by mcourteaux** [comment]
> This reduction is required in the case of `RaggedTensors`. Consider the case:

@Frightera Ah, I have absolutely no experience with ragged tensors. I know what they are, but never worked with them. I won't try to wrap my brain around this.

---

I see that the example given here
https://github.com/keras-team/keras/blob/f9336cc5114b4a9429a242deb264b707379646b7/keras/losses.py#L2121-L2128

is not actually wrong. The y-trues do not sum to one, so this is just a misleading example. The author demonstrates a multi-label scenario with two possible classes per sample. The numbers are chosen somewhat unfortunate. This, in combination with the rule of "a lost must reduce the last dimension", it does all seem reasonable.

Overall, the class weights are just unusable in this multi-label scenario. This I believe can be fixed by also accepting a flat array instead of only a dictionary. But that would require the losses be all aware of the class weights, instead of doing it as the post-processing trick that only works for single-label tasks. Overall, Keras is not really designed with this class-weight thing in mind. A custom loss, or custom training loop is required. Recently, I became a huge fan of custom training loops, as you get much more control of what's happening, and avoid redundant calculations between losses and metrics.
